File: lectures/jvm/src/main/java/pt/isel/pc/lectures/ManualResetEvent.java
package pt.isel.pc.lectures;

public class ManualResetEvent {

    private boolean flag = false;
    private final Object mon = new Object();
    private final NodeLinkedList<Boolean> queue = new NodeLinkedList<>();

    public boolean ewait(long timeout) throws InterruptedException {
        synchronized (mon) {
            if(flag) {
                return true;
            }
            if(Timeouts.noWait(timeout)) {
                return false;
            }
            long t = Timeouts.start(timeout);
            long remaining = Timeouts.remaining(t);
            NodeLinkedList.Node<Boolean> node = queue.push(false);
            while(true) {
                try {
                    mon.wait(remaining);
                }catch(InterruptedException e) {
                    if(node.value != true) {
                        queue.remove(node);
                        throw e;
                    }
                    Thread.currentThread().interrupt();
                    return true;
                }
                if(node.value == true) {
                    return true;
                }
                remaining = Timeouts.remaining(t);
                if(Timeouts.isTimeout(remaining)) {
                    queue.remove(node);
                    return false;
                }
            }
        }
    }

    public void set() {
        synchronized (mon) {
            flag = true;
            while(!queue.isEmpty()) {
                NodeLinkedList.Node<Boolean> node = queue.pull();
                node.value = true;
            }
            mon.notifyAll();
        }
    }

    public void clear() {
        synchronized (mon) {
            flag = false;
        }
    }
}

File: lectures/dotnet/lectures/Isel.Pc.Lectures.Common/ManualResetEvent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Isel.Pc.Lectures.Common
{
    public class ManualResetEvent
    {
        private readonly object mon = new object();
        private volatile bool signaled;
        private volatile int waiters;
        private int setVersion;

        public ManualResetEvent()
        {
            signaled = false;
            waiters = 0;
            setVersion = 0;
        }

        public bool Wait(int timeout)
        {
            if (signaled)
            {
                return true;
            }
            if (!TimeoutInstant.ShouldWait(timeout))
            {
                return false;
            }
            lock (mon)
            {
                waiters += 1;
                Interlocked.MemoryBarrier();
                if(signaled == true)
                {
                    waiters -= 1;
                    return true;
                }
                TimeoutInstant ti = new TimeoutInstant(timeout);
                int myVersion = setVersion;
                while(true)
                {
                    try
                    {
                        Monitor.Wait(mon, ti.Remaining);
                    }
                    catch (Exception)
                    {
                        waiters -= 1;
                        throw;
                    }
                    if (signaled || myVersion != setVersion)
                    {
                        waiters -= 1;
                        return true;
                    }
                    if (ti.IsTimeout)
                    {
                        waiters -= 1;
                        return false;
                    }
                }
            }
        }

        public void Set()
        {
            signaled = true;
            Interlocked.MemoryBarrier();
            if(waiters != 0)
            {
                SyncUtils.EnterUninterruptibly(mon, out bool wasInterrupted);
                setVersion += 1;
                try
                {
                    Monitor.PulseAll(mon);
                }
                finally
                {
                    Monitor.Exit(mon);
                    if(wasInterrupted)
                    {
                        Thread.CurrentThread.Interrupt();
                    }
                }
            }
        }

        public void Reset()
        {
            signaled = false;
        }

    }
}
